File: 785/quick_sort.cpp

#include <iostream>

using namespace std;

const int N = 1e5 + 10;

int q[N], n;

void quick_sort(int left, int right);

int main(){
    scanf("%d", &n);
    for (int i = 0; i < n; i++) scanf("%d", &q[i]);
    quick_sort(0, n - 1);
    for (int i = 0; i < n; i++) printf("%d ", q[i]);
}

void quick_sort(int left, int right) {
    if (left >= right) return;
    int l = left, r = right, pivot = q[l + ((r - l) >> 1)];
    while (true) {
        while (q[l] < pivot) l++;
        while (q[r] > pivot) r--;
        if (l >= r) {
            break;
        }
        swap(q[l++], q[r--]);
    }
    quick_sort(left, r);
    quick_sort(r + 1, right);
}

File: 785/quick_sort.py
def quick_sort(data, left, right):
    if (left >= right):
        return
    l = left
    r = right
    pivot = data[l + ((r - l) >> 1)]
    while True:
        while (data[l] < pivot):
            l += 1
        while (data[r] > pivot):
            r -= 1
        if (l >= r):
            break
        data[l], data[r] = data[r], data[l]
        l += 1
        r -= 1
    quick_sort(data, left, r)
    quick_sort(data, r + 1, right)
    
    
def main():
    n = int(input())
    data = [int(x) for x in input().split()]
    quick_sort(data, 0, n -1)
    print(' '.join(list(map(str, data))))



main()